plutus-core/test/Spec.hs

```diff
@@ -102,13 +105,80 @@ propFlat = property $ do
     prog <- forAllPretty $ runAstGen genProgram
     Hedgehog.tripping prog Flat.flat Flat.unflat
 
--- Generate a random 'Program', pretty-print it, and parse the pretty-printed
+{-| The lexer contains some quite complex regular expressions for literal
+  constants, allowing escape sequences inside quoted strings, among other
+  things.  The lexer returns 'TkLiteralConst' tokens and then individual
+  built-in types interpret these using their own parsing functions via the
+  'Parsable' class.  The following tests check that (A) the lexer/parser can
+  handle the output of the prettyprinter on constants from types in the default
+  universe, and (B) that parsing is left inverse to printing for both constants
+  and programs.  We have unit tests for the unit and boolean types, and property
+  tests for integers, characters, strings, and bytestrings. -}
+
+type DefaultTerm  a = Term TyName Name DefaultUni DefaultFun a
+type DefaultError a = Error DefaultUni DefaultFun a
+
+parseTm :: BSL.ByteString -> Either (DefaultError AlexPosn) (DefaultTerm AlexPosn)
+parseTm = runQuote . runExceptT . parseTerm @(DefaultError AlexPosn)
+
+reprint :: PrettyPlc a => a -> BSL.ByteString
+reprint = BSL.fromStrict . encodeUtf8 . displayPlcDef
+
+{-| Test that the lexer/parser can successfully consume the output from the
+   prettyprinter for the unit and boolean types.  We use a unit test here
+   because there are only three possiblities (@()@, @false@, and @true@). -}
+testLexConstants :: Assertion
+testLexConstants =
+    mapM_ (\t -> (fmap void . parseTm . reprint $ t) @?= Right t) smallConsts
+        where
+          smallConsts :: [DefaultTerm ()]
+          smallConsts =
+              [ mkConstant () ()
+              , mkConstant () False
+              , mkConstant () True
+              ]
+
+{-| Generate constant terms of type @integer@, @char@, @string@, and @bytestring@.
+  The lexer should be able to consume escape sequences in characters and strings;
+  both standard ASCII escape sequences and Unicode ones.  Hedgehog has generators
+  for both of these, but the Unicode one essentially never generates anything readable:
+  all of the output looks like '\857811'.  For this reason we have separate generators
+  for Unicode characters and Latin-1 ones (characters 0-255, including standard
+  ASCII from 0-127); there is also a generator for UTF8-encoded Unicode. -}
+
+-- TODO: Language.PlutusCore.Generators.AST has a 'genConstant' function, but it only
+-- generates integers and UTF8-encoded Unicode strings. Should we replace that with
+-- this function?
```

Comment:
Not sure, perhaps @effectfully has an opinion?

Reply:
Yes, but I thought he was on holiday.  I could open another issue to return to this once he's back.

Reply:
> Should we replace that with this function?

Seems reasonable to me.

> Yes, but I thought he was on holiday. I could open another issue to return to this once he's back.

It doesn't take long to answer a question (sometimes with "dunno, I'll think about it once I'm back"), so feel free to ping me.

Reply:
OK, maybe I'll stick that in `AST.hs` but still keep separate tests for constants to make sure we get good coverage.  I'll think a bit more about what to do for `()` and `bool`; maybe stick them in there with low frequencies.

Reply:
I've now done that.